### PR TITLE
Dataplane policy states and events API swagger 

### PR DESCRIPTION
**Pull request by NarineM**
### Latest improvements:
Added data plane events and states API under version 2019-10-01. Added samples. 
<br>

@pilor

### CODE REVIEW

**Comment by ravbhatnagar** [comment]
@NarineM - can you please set something up with armapireview to go over this? It will be much quicker in person since this is a new type.

**Comment by NarineM** [comment]
@ravbhatnagar  ok, thanks, I will setup a meeting.

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_QueryResourceScopeExpandComponentsGroupByWithAggregate.json
Comment (raych1) [inline specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_QueryResourceScopeExpandComponentsGroupByWithAggregate.json]: Property of `totalActions` is not defined in the swagger [here](https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json#L755)

**Comment by raych1** [comment]
@NarineM , there're two additional property not defined in swagger which causes the model validation failure. One I have already commented inline. another one is  in `specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyStates_QueryResourceScopeExpandComponentsGroupByWithAggregate.json   jsonPath: '$.value[0].components.components[0].count'` . Please check the build log for details.

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_QueryResourceScopeExpandComponentsGroupByWithAggregate.json
Comment (NarineM) [reply specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_QueryResourceScopeExpandComponentsGroupByWithAggregate.json]: The issue must be something else, because there are quite a few other json files that are written in the same manner, where they take custom property names from the  request URL and are not defined in the schema, for some reason model validation is passing for these root properties but not for nested properties(that are in my PR). Here is an example of similar json for which validation is passing:
https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_FilterAndAggregateOnly.json

If you look in this example NumDenyEvents property is not defined in the swagger for this API, but validation is passing.

**Comment by NarineM** [comment]
@raych1, the root cause is something else please see this json for which validation is passing where it contains a property NumDenyEvents  which is not defined in events API swagger at: https://github.com/NarineM/azure-rest-api-specs/blob/narinem-dataplane-swagger/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json

Here is the json with similar implementation but Model validation is passing for this one:
{
  "parameters": {
    "policyEventsResource": "default",
    "subscriptionId": "fffedd8f-ffff-fffd-fffd-fffed2f84852",
    "api-version": "2019-10-01",
    "$from": "2018-02-05T18:00:00Z",
    "$filter": "PolicyDefinitionAction eq 'deny'",
    "$apply": "aggregate($count as NumDenyEvents)"
  },
  "responses": {
    "200": {
      "body": {
        "@odata.context": "https://management.azure.com/subscriptions/fffedd8f-ffff-fffd-fffd-fffed2f84852/providers/Microsoft.PolicyInsights/policyEvents/$metadata#default",
        "@odata.count": 1,
        "value": [
          {
            "@odata.id": null,
            "@odata.context": "https://management.azure.com/subscriptions/fffedd8f-ffff-fffd-fffd-fffed2f84852/providers/Microsoft.PolicyInsights/policyEvents/$metadata#default/$entity",
            "NumDenyEvents": 40
          }
        ]
      }
    }
  }
}

**Comment by raych1** [comment]
Thanks @NarineM for raising this. I will double check.

**Comment by raych1** [comment]
@NarineM , you need to set `additionalProperties` in the model if it will have undefined properties. Please see [example](https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json#L748)

**Comment by NarineM** [comment]
@raych1 thanks that worked. Now I am getting breaking change violation on "components" property being published under same API version. Here is the error from validating breaking change : https://dev.azure.com/azure-sdk/public/_build/results?buildId=303511&view=logs&j=08b5038d-1e84-5877-88da-859df8312b69&t=c1978ecd-fc8a-5291-6d7a-7dd459d8b50f

If I need this change be published for the same API , how do I bypass the breaking change validation error?

**Comment by raych1** [comment]
@NarineM , prettier failure is about JSON format issue. Please follow https://aka.ms/AA6h31t to fix the issue. 
Introduce new property in stable version API is considered as breaking change. Can that be published with new version? If not, it needs to request approval on force merge.

**Comment by NarineM** [comment]
@raych1 I have ran Prettier check as it is documented at the link you provided  , but can not figure out what is wrong. Looking at the log looks like the tool is not even running. This is what I get when I run the tool:
PS C:\git\azure-rest-api-specs> npm run prettier -- --write "specification/policyinsights/**/*.json"

> azure-rest-api-specs-tests@0.1.0 prettier C:\git\azure-rest-api-specs
> prettier "--write" "specification/policyinsights/**/*.json"

[error] Unable to expand glob patterns: specification/policyinsights/**/*.json !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
[error] Cannot find module 'C:\git\azure-rest-api-specs\scripts\prettier-swagger-plugin' from 'C:\git\azure-rest-api-specs'
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! azure-rest-api-specs-tests@0.1.0 prettier: `prettier "--write" "specification/policyinsights/**/*.json"`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the azure-rest-api-specs-tests@0.1.0 prettier script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\narinem\AppData\Roaming\npm-cache\_logs\2020-03-17T18_49_42_976Z-debug.log

**Comment by NarineM** [comment]
@raych1 I have validated that these json files are correct based on RFC 8259. Using tools like https://jsonformatter.curiousconcept.com/
so I am not quite sure whether it is an issue with the tool or the files?

**Comment by anthony-c-martin** [comment]
I'm happy from an ARM perspective as we're really just adding filtering options to an existing API. @pilor is going to review from a Policy perspective.

**Comment by raych1** [comment]
> @raych1 I have validated that these json files are correct based on RFC 8259. Using tools like https://jsonformatter.curiousconcept.com/
> so I am not quite sure whether it is an issue with the tool or the files?

The tool only detects the style issue in JSON. Can you make sure the script `C:\git\azure-rest-api-specs\scripts\prettier-swagger-plugin` exist locally?

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyStates.json
Comment (pilor) [inline specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyStates.json]: >By default this is not populated. [](start = 63, length = 33)

Instead of saying this can you specify when this would be populated?

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (pilor) [inline specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: >By default this is not populated. [](start = 52, length = 33)

Same comment as policyStates.  Please specify when this is populated

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (pilor) [inline specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: >"tenantId" [](start = 2, length = 10)

This can be different than the root policyEvent tenantId?

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (pilor) [inline specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: > to expand policyEvaluationDetails, use $expand=policyEvaluationDetails [](start = 63, length = 71)

policyEvaluationDetails doesn't exist in policyEvents right?

**Review by pilor — CHANGES_REQUESTED** [review]
A few minor comments

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (NarineM) [reply specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: Yes, the user for keyvault can be different from the user operating over certificates that belong to given keyvault.

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (NarineM) [reply specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: Ah, yes, thanks for catching it. Fixed.

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json
Comment (NarineM) [reply specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json]: Sure, fixing it now.

File: specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyStates.json
Comment (NarineM) [reply specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyStates.json]: fixed.

**Comment by NarineM** [comment]
@raych1 yes, the swagger prettier script exists in the path, I have pulled the entire repo that sits on github.

**Comment by NarineM** [comment]
@pilor Just pushed a new iteration that addresses your comments. I can see this warning from cli-extension, do you know what it complains about?:
"azure-cli-extensions
No readme.md specification configuration files were found that are associated with the files modified in this pull request, or swagger_to_sdk section in readme.md is not configured "

**Comment by pilor** [comment]
> @pilor Just pushed a new iteration that addresses your comments. I can see this warning from cli-extension, do you know what it complains about?:
> "azure-cli-extensions
> No readme.md specification configuration files were found that are associated with the files modified in this pull request, or swagger_to_sdk section in readme.md is not configured "

No, I don't know the specifics on that but I have seen it on my own PRs which went through just fine.  The SDK reviewer is a better contact.

**Review by pilor — APPROVED** [review]

**Comment by NarineM** [comment]
@raych1 Can you help with this new error from SDK automation log?, it says branch changes need to be squished. Looks like something with the tool.

**Comment by raych1** [comment]
@NullMDR , can you help with the SDK automation failures?

**Comment by NarineM** [comment]
@raych1  thanks. So far all PR merge validation issues have been resolved except prettier check. This PR can be merged unless you tell me the prettier check issue is blocking. As of now I can not figure out why the tool is not running or what it complains about.

**Comment by NarineM** [comment]
Please go ahead with merging this PR when you get a chance if you do not find anything with prettier check.

**Comment by raych1** [comment]
@NarineM, prettierCheck is required to pass. @NullMDR , can you help Narine, he failed to run prettier?

**Comment by NullMDR** [comment]
@NarineM I've fixed your prettier failure by running 
`
npm run prettier -- --write "specification/policyinsights/**/*.json"
`
But please fix the merge conflict.

**Comment by NarineM** [comment]
@NullMDR, thank you. I have resolved merge conflics. The only failure now remaining is related to api version (under breaking change catergory).  As we discussed earlier this change needs to be published for the same api version 2019-10-01. Please go ahead with merging this PR.

**Comment by NullMDR** [comment]
@jianyexi Why adding property in response will be considered breaking change?

**Comment by NarineM** [comment]
@NullMDR  It is not breaking change because this property is only appended when the client specifies it in request URL with $expand clause. This functionality is available in the current version of API on the RP side, so it should not be published under new API version.

**Comment by NarineM** [comment]
@NullMDR , can you go ahead with merging this PR? Let me know what is blocking if any. thanks.

**Comment by ArcturusZhang** [comment]
> @NullMDR It is not breaking change because this property is only appended when the client specifies it in request URL with $expand clause. This functionality is available in the current version of API on the RP side, so it should be published under the same API version 2019-10-01 and not a new version.

I think this is reasonable. 

Hi @akning-ms would you please have us merge this PR? This breaking change should be acceptable based on the author's comments

**Pull request merged**
